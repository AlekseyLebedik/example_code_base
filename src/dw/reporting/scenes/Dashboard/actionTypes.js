export const FETCH_FRANCHISE_DATA = 'REPORTING.FETCH_FRANCHISE_DATA';
export const FETCH_FRANCHISE_DATA_SUCCESS =
  'REPORTING.FETCH_FRANCHISE_DATA_SUCCESS';
export const FETCH_FRANCHISE_DATA_FAIL = 'REPORTING.FETCH_FRANCHISE_DATA_FAIL';
