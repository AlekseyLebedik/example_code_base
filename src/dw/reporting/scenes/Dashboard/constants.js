const WEEKS_NO = 5;

export { WEEKS_NO };
