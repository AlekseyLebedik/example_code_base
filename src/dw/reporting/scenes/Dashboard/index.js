import ReportingContainer from './container';

export default ReportingContainer;
