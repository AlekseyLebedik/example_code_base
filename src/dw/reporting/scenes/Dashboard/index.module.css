.container {
  padding: 30px;
  color: rgba(0, 0, 0, 0.5);
}

.container a {
  color: rgba(0, 0, 0, 0.5);
}

.container :global(.hoverable) {
  visibility: hidden;
  opacity: 0;
}

@media (max-width: 768px) {
  .container {
    padding: 0;
  }
}
