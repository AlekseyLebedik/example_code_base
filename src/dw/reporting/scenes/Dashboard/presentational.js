import React from 'react';
import PropTypes from 'prop-types';
import { AutoSizer } from 'react-virtualized';
import { Row, Col } from 'antd';

import TrendTable from './components/TrendTable';
import PlatformsTable from './components/PlatformsTable';
import TrendCharts from './components/TrendCharts';

import styles from './index.module.css';

const DashboardStateless = ({ franchise, franchiseData, location }) => (
  <div className={styles.container}>
    <TrendTable franchiseData={franchiseData} baseUrl={location.pathname} />
    <Row>
      <Col sm={24} md={12}>
        <TrendCharts
          franchiseData={franchiseData}
          baseUrl={location.pathname}
        />
      </Col>
      <Col sm={24} md={12}>
        <AutoSizer disableHeight>
          {({ width }) => (
            <PlatformsTable
              key="1"
              franchise={franchise}
              baseUrl={location.pathname}
              availableWidth={width}
            />
          )}
        </AutoSizer>
      </Col>
    </Row>
  </div>
);

DashboardStateless.propTypes = {
  franchise: PropTypes.object,
  franchiseData: PropTypes.object,
  location: PropTypes.object.isRequired,
};

DashboardStateless.defaultProps = {
  franchise: null,
  franchiseData: {},
};

export default DashboardStateless;
