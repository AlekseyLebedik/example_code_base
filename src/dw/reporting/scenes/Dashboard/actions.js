import moment from 'moment';

import * as AT from './actionTypes';

const fetchFranchiseData = id => {
  const end = Date.now();
  const start = moment(end).subtract(5, 'weeks');
  return {
    type: AT.FETCH_FRANCHISE_DATA,
    id,
    start: start.valueOf(),
    end,
  };
};

const fetchFranchiseDataSuccess = ({ id, data, statName }) => ({
  type: AT.FETCH_FRANCHISE_DATA_SUCCESS,
  id,
  data,
  statName,
});

const fetchFranchiseDataFailed = () => ({
  type: AT.FETCH_FRANCHISE_DATA_FAIL,
});

export {
  fetchFranchiseData,
  fetchFranchiseDataSuccess,
  fetchFranchiseDataFailed,
};
