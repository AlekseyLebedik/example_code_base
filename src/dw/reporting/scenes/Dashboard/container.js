import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import Error404 from 'dw/core/components/Error404';
import {
  selectedFranchiseSelector,
  franchiseDataSelector,
} from 'dw/reporting/selectors';

import { fetchFranchiseData } from './actions';

import DashboardStateless from './presentational';

const stateToProps = (state, props) => ({
  franchise: selectedFranchiseSelector(state, props),
  franchiseData: franchiseDataSelector(state),
});

const dispatchToProps = {
  fetchFranchiseData,
};

class Dashboard extends Component {
  state = {};

  static getDerivedStateFromProps(props) {
    if (props.match.params.franchiseId !== props.franchiseData.id) {
      props.fetchFranchiseData(props.match.params.franchiseId);
    }
    return null;
  }

  render() {
    if (this.props.franchise !== undefined) {
      return <DashboardStateless {...this.props} />;
    }
    const redirectTo = this.props.match.path.split(':franchiseId')[0];
    return <Error404 redirectTo={redirectTo} />;
  }
}

Dashboard.propTypes = {
  franchise: PropTypes.object,
  match: PropTypes.object.isRequired,
  franchiseData: PropTypes.object,
  fetchFranchiseData: PropTypes.func.isRequired,
};

Dashboard.defaultProps = {
  franchise: undefined,
  franchiseData: {},
};

export default connect(stateToProps, dispatchToProps)(Dashboard);
