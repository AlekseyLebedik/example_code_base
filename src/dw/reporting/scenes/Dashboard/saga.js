import { all, spawn, call, put, takeLatest } from 'redux-saga/effects';

import { getStatData as apiGetStatData } from 'dw/devzone/services/graph';
import { STAT_NAMES } from 'dw/reporting/constants';
import * as Actions from './actions';
import { FETCH_FRANCHISE_DATA } from './actionTypes';

export function* fetchStats(APICall, statName, params) {
  const { id, start, end } = params;
  const source = {
    resource: `franchises/${id}/stats`,
  };
  try {
    const { data } = yield call(APICall, statName, source, start, end);
    const result = {
      id,
      statName,
      data: data.series.find(s => s.id === 'total') || { data: [] },
      start,
      end,
    };
    yield put(Actions.fetchFranchiseDataSuccess(result, params.append));
  } catch (e) {
    yield put(Actions.fetchFranchiseDataFailed(e));
  }
}

export function* fetchFranchiseData(action) {
  try {
    yield all(
      STAT_NAMES.map(statName =>
        spawn(fetchStats, apiGetStatData, statName, action)
      )
    );
  } catch (e) {
    yield put(Actions.fetchFranchiseDataFailed(e));
  }
}

function* saga() {
  yield takeLatest(FETCH_FRANCHISE_DATA, fetchFranchiseData);
}

export default saga;
