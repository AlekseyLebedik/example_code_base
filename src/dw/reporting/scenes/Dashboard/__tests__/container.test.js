import React from 'react';
import { shallowUntilTarget } from 'dw/test-utils';
import Error404 from 'dw/core/components/Error404';
import createStore from 'dw/reporting/store';
import * as actions from 'dw/reporting/actions';
import { franchises as franchisesMock } from 'dw/reporting/__mocks__/franchises';
import Dashboard from '../container';

describe('Dashboard container', () => {
  let store;
  beforeEach(() => {
    // eslint-disable-next-line
    store = createStore().store;
  });
  it('no franchises', () => {
    store.dispatch(actions.fetchFranchisesSucceed([]));
    const props = {
      match: { path: '/reporting', params: { franchiseId: 1 } },
      store,
    };
    const wrapper = shallowUntilTarget(<Dashboard {...props} />, Error404);
    expect(wrapper).toHaveLength(1);
  });

  it('renders TrendTable, TrendCharts and PlatformsTable', () => {
    store.dispatch(actions.fetchFranchisesSucceed(franchisesMock));
    const props = {
      history: {},
      location: { pathname: '/reporting/1' },
      match: {
        path: '/reporting/:franchiseId',
        params: { franchiseId: 1 },
      },
      store,
    };
    const wrapper = shallowUntilTarget(
      <Dashboard {...props} />,
      'DashboardStateless'
    );
    expect(wrapper).toMatchSnapshot();
    const sizer = wrapper.find('AutoSizer').shallow();
    expect(sizer).toMatchSnapshot();
  });
});
