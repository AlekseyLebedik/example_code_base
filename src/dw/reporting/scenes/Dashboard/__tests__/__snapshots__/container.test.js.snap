// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard container renders TrendTable, TrendCharts and PlatformsTable 1`] = `
<div
  className="container"
>
  <WithWidth(TrendTableComponent)
    baseUrl="/reporting/1"
    franchiseData={Object {}}
  />
  <Row
    gutter={0}
  >
    <Col
      md={12}
      sm={24}
    >
      <WithWidth(TrendChartsComponent)
        baseUrl="/reporting/1"
        franchiseData={Object {}}
      />
    </Col>
    <Col
      md={12}
      sm={24}
    >
      <AutoSizer
        disableHeight={true}
        disableWidth={false}
        onResize={[Function]}
        style={Object {}}
      >
        <Component />
      </AutoSizer>
    </Col>
  </Row>
</div>
`;

exports[`Dashboard container renders TrendTable, TrendCharts and PlatformsTable 2`] = `
<div
  style={
    Object {
      "overflow": "visible",
      "width": 0,
    }
  }
>
  <WithWidth(Connect(HigherOrderComponent))
    availableWidth={0}
    baseUrl="/reporting/1"
    franchise={
      Object {
        "id": 1,
        "name": "Simple Franchise",
        "projects": Array [
          Object {
            "id": 1,
            "name": "Simple Project",
            "platforms": Array [
              Object {
                "platform": "PS4",
              },
              Object {
                "platform": "Xbox One",
              },
              Object {
                "platform": "PC",
              },
              Object {
                "platform": "Wii",
              },
            ],
          },
        ],
      }
    }
    key="1"
  />
</div>
`;
